doc/developer/design/20230110_window_functions.md

```diff
@@ -0,0 +1,331 @@
+
+For this to perform well, we are assuming that groups are small.
+This is not an unreasonable assumption, because a group is identified here by a value of the PARTITION BY expression + a value of the ORDER BY expression.
+Also note that if there is no ORDER BY, then groups might be large, but in this case we don’t employ Prefix Sum, but we transform away the window functions to grouped aggregation + self-join (as noted above).
+
+Alternatively, we could change Prefix Sum so that it correctly handles duplicate indexes.
+
+### Parallelism
+
+DD's Prefix Sum should be data-parallel even inside a window partition. (It’s similar to [a Fenwick tree](https://en.wikipedia.org/wiki/Fenwick_tree), with sums maintained over power-of-2 sized intervals, from which you can compute a prefix sum by putting together LogN intervals.) TODO: But I wasn't able to actually observe a speedup in a simple test when adding cores, so we should investigate what’s going on with parallelization. There was probably just some technical issue, because all operations in the Prefix Sum implementation look parallelizable.
+
+### Types
+
+We'll have to generalize DD's Prefix Sum to orderings over types other than a single unsigned integer, which is currently hardcoded in the code that forms the intervals. We’ll map other types to a single unsigned integer. Importantly, this mapping should *preserve the ordering* of the type:
+
+- Signed integer types are fine, we just need to fiddle with the sign to map them to an unsigned int in a way that preserves the ordering.
+- Date/Time types are just a few integers. We’ll concatenate their bits.
+- I don’t know how to handle strings, so these are out of scope for now. (Not seen in user queries yet.)
+
+## Alternatives
+
+### Where to put the idiom recognition?
+
+We could do the idiom recognition in the rendering instead of the MIR-to-LIR lowering. However, the lowering seems to be a more natural place for it:
+- We shouldn't have conditional code in the rendering, and this idiom recognition will be a giant piece of conditional code,
+- We want EXPLAIN PHYSICAL PLAN to show how we'll execute a window function.
+- We need to know the type of the ORDER BY columns, which we don't know in LIR. (Although we could add extra type info to window function calls.)
+
+### Use WITH MUTUALLY RECURSIVE
+
+We are [in the process of adding support for recursive SQL queries](https://github.com/MaterializeInc/materialize/issues/17012). We could implement prefix sum using that instead of custom rendering code. In this case, all the window function handling could happen in the HIR-to-MIR lowering, by translating to a MIR expression that has `LetRec`.
+
+### Representing window functions in each of the IRs
```

Comment:
This section discusses what options we have, but it doesn't discuss trade-offs. It's not a high-priority item, but I think that as we're working on this, we should make an explicit decision which representation we select and for what reasons.

Reply:
I added more discussion on trade-offs of the different representations.